sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/RpcQosImpl.java

```diff
@@ -0,0 +1,695 @@
+      long nextBatchMaxCount =  options.getBatchTargetLatency().getMillis() /  recentMeanLatency;
+      return Math.toIntExact(nextBatchMaxCount);
+    }
+
+  }
+
+  /**
+   * An implementation of client-side adaptive throttling. See
+   * https://sre.google/sre-book/handling-overload/#client-side-throttling-a7sYUg
+   * for a full discussion of the use case and algorithm applied.
+   */
+  private final class AdaptiveThrottler {
+    private final MovingFunction successfulRequestsMovingFunction;
+    private final MovingFunction failedRequestsMovingFunction;
+    private final MovingFunction allRequestsMovingFunction;
+
+    private AdaptiveThrottler() {
+      allRequestsMovingFunction = createMovingFunction(options.getSamplePeriod(), options.getSamplePeriodBucketSize());
+      successfulRequestsMovingFunction = createMovingFunction(options.getSamplePeriod(), options.getSamplePeriodBucketSize());
+      failedRequestsMovingFunction = createMovingFunction(options.getSamplePeriod(), options.getSamplePeriodBucketSize());
+    }
+
+    private Duration shouldThrottleRequest(Instant instantSinceEpoch) {
+      double delayProbability = throttlingProbability(instantSinceEpoch);
+
+      return (random.nextDouble() < delayProbability) ? options.getThrottleDuration() : Duration.ZERO;
+    }
+
+    private void recordStartRequest(Instant instantSinceEpoch) {
+      allRequestsMovingFunction.add(instantSinceEpoch.getMillis(), 1);
+    }
+
+    private void recordSuccessfulRequest(Instant instantSinceEpoch) {
+      successfulRequestsMovingFunction.add(instantSinceEpoch.getMillis(), 1);
+    }
+
+    private void recordFailedRequest(Instant instantSinceEpoch) {
+      failedRequestsMovingFunction.add(instantSinceEpoch.getMillis(), 1);
+    }
+
+    /**
+     * Implementation of the formula from https://sre.google/sre-book/handling-overload/#eq2101
+     */
+    private double throttlingProbability(Instant instantSinceEpoch) {
+      if (!allRequestsMovingFunction.isSignificant()) {
+        return 0;
+      }
+      long nowMsSinceEpoch = instantSinceEpoch.getMillis();
+      long allRequestsCount = allRequestsMovingFunction.get(nowMsSinceEpoch);
+      long successfulRequestsCount = successfulRequestsMovingFunction.get(nowMsSinceEpoch);
+
+      double overloadMaxCount = options.getOverloadRatio() * successfulRequestsCount;
+      double overloadUsage = allRequestsCount - overloadMaxCount;
+
+      double calcProbability = overloadUsage / (allRequestsCount + MIN_REQUESTS);
+      return Math.max(0, calcProbability);
+    }
+  }
+
+  /**
+   * An implementation providing the 500/50/5 ramp up strategy recommended by <a
+   * href="https://cloud.google.com/firestore/docs/best-practices#ramping_up_traffic">Ramping up
+   * traffic</a>
+   */
+  @VisibleForTesting
+  static final class WriteRampUp {
+    private static final Duration RAMP_UP_INTERVAL = Duration.standardMinutes(5);
+    private final int baseMax;
```

Comment:
`initialWriteBudget` or `baseBatchBudget`?

Reply:
I'll change it to `baseBatchBudget`